Add matchWhitespaces option to cy.contains.

sainthkh opened the pull request:
- Closes #6405

### User facing changelog

You can opt out white space collapsing in `cy.contains`.

### Additional details

#### Why was this change necessary?

* Sometimes, you might want to opt out white space collapsing.
* Support opting out white space collapsing when `pre`, `pre-line`, `pre-wrap`, `break-spaces` of `white-space` CSS style is sued. 

#### What is affected by this change?

* `matchWhitespaces` option is added to `cy.contains`

#### Any implementation details to explain?

N/A

### How has the user experience changed?

* Can opt out with `matchWhitespaces` option like this:
```js
cy.get('#btn').contains('  name    age   ', { matchWhitespaces: true })`
```
* Opt out collapsing when `pre`, `pre-line`, `pre-wrap`, `break-spaces` are used for `white-space` style. 

### Note

`white-space: pre-line` collapses white spaces. But the implementation of Chrome and Firefox is a bit different. Chrome collapses the white spaces at the end of a line to a single space, while Firefox removes them.

**Chrome:**

![Peek 2020-02-11 17-32](https://user-images.githubusercontent.com/8130013/74222097-00117a80-4cf7-11ea-9560-b6c91ba7bf74.gif)

**Firefox:**

![Peek 2020-02-11 17-33](https://user-images.githubusercontent.com/8130013/74222121-0d2e6980-4cf7-11ea-9c62-f74f4cc8b532.gif)

It seems that the behavior of Firefox is correct because those trailing spaces are collapsed in `normal` blocks. 

Below is the code used for test. 

```html
<h2>white-space: pre-line:</h2>
<p class="e">
         This       			 is some text. This is some text. This is some text.

This is some text. This is some text. This is some text.
This is some text. This is some text. This is some text.       
This is some text. This is some text. This is some text.           
</p>
```

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [N/A] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

Note: Documentation and index.d.ts update will be done when this behavior is confirmed.

c32hedge commented:
I added a relevant comment on the associated issue: https://github.com/cypress-io/cypress/issues/6405#issuecomment-584633738

chrisbreiding commented on packages/driver/src/cy/commands/querying.js:
```diff
@@ -413,7 +414,7 @@ module.exports = (Commands, Cypress, cy) => {
       }
 
-      _.defaults(options, { log: true, matchCase: true })
+      _.defaults(options, { log: true, matchCase: true, matchWhitespaces: false })
```
Let's change the option to `matchWhitespace` (no `s` at the end).

chrisbreiding commented on packages/driver/src/cy/commands/querying.js:
```diff
@@ -486,11 +487,24 @@ module.exports = (Commands, Cypress, cy) => {
+          })
+
+          return lines.join('\n')
+        }
```
I think we can't add this behavior regarding the `white-space` CSS property without this being a breaking change. It should only take into account the `matchWhitespace` property. Maybe it makes sense to also handle `pre-line` specially, but again, only if `matchWhitespace: true`.

c32hedge replied:
I'm curious if everything will "just work", removing the need to check for `white-space`, or even having a `matchWhitespace` option, if the calls to `normalizeWhitespaces(elem)` are just replaced with `let testText = elem.innerText`. See [Node.textContent#Differences from innerText](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#Differences_from_innerText) and [HTMLElement.innerText](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/innerText), which states:

> The innerText property of the HTMLElement interface represents the "rendered" text content of a node and its descendants. **As a getter, it approximates the text the user would get if they highlighted the contents of the element with the cursor and then copied it to the clipboard.**

The current code for `normalizeWhitespaces` gives `textContent` higher precedent than `innerText`. I also found that Cypress currently includes the text for any inline styles in the `testText`, which is consistent with the documentation for `textContent.

Bkucera commented:
I think it's confusing that the parent querying command `.contains(...)` has different logic than the assertion `expect(el).contains(...)`

brian-mann commented:
I'm doing review on this - from where we implemented the `cy.contains()` breaking change in 4.0 to the suggestions here.

I'll leave my feedback here in a little bit...

c32hedge commented:
Having had time to think about this since I entered the original issue yesterday, I'm not sure having Cypress pay attention to the css `white-space` property is actually what I want. I attempted to clarify in [another comment on the original issue](https://github.com/cypress-io/cypress/issues/6405#issuecomment-584710192) and included a `cypress-test-tiny` repo with some tests.

Sorry if the comments on both issues are redundant--the PR seems to be getting more attention so I wanted to make sure the updated feedback isn't missed.

sainthkh commented on packages/driver/test/cypress/integration/commands/querying_spec.js:
```diff
@@ -1567,13 +1567,13 @@ describe('src/cy/commands/querying', () => {
       cy.$$('<script>// some-script-content </script>').appendTo(cy.$$('body'))
 
-      cy.contains('some-script-content').should('not.match', 'script')
+      cy.contains('some-script-content').should('not.exist')
```
As we use `innerText` now, contains cannot find the tag at all because text in script tag is invisible

sainthkh commented:
Before I say anything, I want to apologize. I was too careless about changing the public API. I borrowed the idea from @c32hedge and changed `normalizeWhitespace`. It fixes all `matchWhitespace` related problems. 

But the problem is that it is another breaking change. 

I'm waiting for the feedback.

sainthkh commented:
Rebased and fixed failures.

[As for Firefox failure](https://circleci.com/gh/cypress-io/cypress/267717?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link), it was caused by a bug in Firefox. 

[Check this bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=1622216)

sainthkh commented:
@jennifer-shehane Shouldn't we resume the discussion about this PR, too?

jennifer-shehane commented:
I don't think this is part of our current 5.0 release plan. Might want to get @JessicaSachs eyes on this PR in general. 👀

sainthkh commented:
Closing because things have changed a lot.
